### ISSUE TITLE
Don’t Re-Create API if not needed

### ISSUE THREAD

**Issue by cwiechmann** [Axway-API-Management-Plus/apim-cli#95]
**User story**
Currently, for most of the API properties (e.g. Policies, Security) the API is re-created including the Backend-API, when the API is in published state.
This is technically not required, as it would be sufficiently unpublish the API, apply all changes and re-publish.

- This will improve the performance as a re-import of the BE not done
- The FE-API ID stays, which might be handy

**Additional context**
Most the CLI-Code is well prepared to support this behavior.
Basically, an additional flag: requiresReImport is needed and handled in APIChangeState

**Comment by lacg** [comment]
Hi Chris, I am also having the same behaviour with the pending state. Is that right?
So, the work around is to use the CLI tool to unpublish the API, do the changes and then publish again for now?
Thanks, Luiz

**Comment by asolipeta11** [comment]
Hi Christoph,

I have a published API which I unpublished, then tried using the "apim api import" command to update the existing API with a new vhost defined in the config JSON. 
apim-cli still seems to be creating a duplicate API regardless of the existing API's state.

Thanks,
Anvesh

**Comment by cwiechmann** [comment]
> So, the work around is to use the CLI tool to unpublish the API, do the changes and then publish again for now?

No, you don’t need to unpublish the API manually. This happens by the CLI automatically transparent to you.
If the CLI realizes an attribute cannot be changes in a certain API-State it brings the API to the required state to do the change. Unfortunately it actually does more than needed (Re-Creating the API), but the end result is the same.
This issue is about that most of the time the API is re-created which is not needed. This enhancement will improve this behavior.

**Comment by cwiechmann** [comment]
> I have a published API which I unpublished, then tried using the "apim api import" command to update the existing API with a new vhost defined in the config JSON.
> apim-cli still seems to be creating a duplicate API regardless of the existing API's state.
This behavior is expected, as the V-Host is a criteria to identify an existing API.
See here for more details:
https://github.com/Axway-API-Management-Plus/apim-cli/wiki/2.1.2-API-Configuration-file#how-the-actual-api-is-looked-upped
Perhaps that is a good candidate for the api change command. What do you think?

**Comment by lacg** [comment]
> > So, the work around is to use the CLI tool to unpublish the API, do the changes and then publish again for now?
> 
> No, you don’t need to unpublish the API manually. This happens by the CLI automatically transparent to you.
> If the CLI realizes an attribute cannot be changes in a certain API-State it brings the API to the required state to do the change. Unfortunately it actually does more than needed (Re-Creating the API), but the end result is the same.
> This issue is about that most of the time the API is re-created which is not needed. This enhancement will improve this behavior.

Hi Chris, so don't we have a way to use the CLI tool to approve an API? Right now, it is duplicating the API without the path.
Is this the same issue, a new issue, or not an issue at all, but a feature?
Thanks, Luiz

**Comment by cwiechmann** [comment]
> Hi Chris, so don't we have a way to use the CLI tool to approve an API? Right now, it is duplicating the API without the path.
> Is this the same issue, a new issue, or not an issue at all, but a feature?

There is currently no interactive command to approve an API which is in pending state. So, that would be a new feature. If you have a need for that please create a new issue.

**Comment by lacg** [comment]
> > Hi Chris, so don't we have a way to use the CLI tool to approve an API? Right now, it is duplicating the API without the path.
> > Is this the same issue, a new issue, or not an issue at all, but a feature?
> 
> There is currently no interactive command to approve an API which is in pending state. So, that would be a new feature. If you have a need for that please create a new issue.

Thank you Chris, for the quick reply, I just created the [enhancement #97](https://github.com/Axway-API-Management-Plus/apim-cli/issues/97).
Luiz

**Comment by gwandwingor** [comment]
Hi @lacg, 

FYI, Starting from Sept20 release, Org Admin users will be empowered to Publish/Unpublish APIs without the need for an API Admin to approve. This could help reducing the number of APIs in "Pending" state.

Link to Sept20 release note : [https://community.axway.com/s/article/Axway-AMPLIFY-API-Management-September-2020-Update](https://community.axway.com/s/article/Axway-AMPLIFY-API-Management-September-2020-Update)

**Comment by lacg** [comment]
> Hi @lacg,
> 
> FYI, Starting from Sept20 release, Org Admin users will be empowered to Publish/Unpublish APIs without the need for an API Admin to approve. This could help reducing the number of APIs in "Pending" state.
> 
> Link to Sept20 release note : https://community.axway.com/s/article/Axway-AMPLIFY-API-Management-September-2020-Update

Thank you, I forgot about this one and I will say to them as one more reason to proceed further to the September release.

**Comment by cwiechmann** [comment]
I have discussed this requirement once more internally and one of the reasons, why the APIM-CLI is Re-Creating APIs with every change, once it's published is: Zero-Down-Time. It basically follows the process designed by the API-Manager to support Zero-Downtime updates of APIs. 
Anyway, we concluded it's not required to re-import the entire API (BE-API & FE-API), just to configure a certain setting of the Frontend-API, hence we have decided to make this new behavior the default: Don't Re-Create API if not needed. 

However, I have added a new parameter allowing customers to control the behavior:
```
# If a published API can be updated without the need to re-import the 
# backend API (e.g. changing a policy, tags, etc.), the frontend API is unpublished, 
# changed and then republished. 
# This process only takes a few seconds, but technically leads to a short downtime for consumers. 
# The advantage is that in this mode the API-ID and used runtime quotas of applications are preserved.
# This is the default behavior since version 1.3.0.
#
# You can set the switch zeroDowntimeUpdate to true to update APIs with zero downtime as before. 
# That means that when updating a published API, the CLI always creates a completely new API, 
# configures it according to the old API and finally clients are switched to the new API. 
# This new API then has a new ID and runtime quotas are lost. The configured quotas are 
# transferred to the new API. 
# zeroDowntimeUpdate=true
```

I'll make this change very clear in the release note of version 1.3.0.
